core/spying/type-matcher.ts

```diff
@@ -1,30 +1,124 @@
-export class TypeMatcher {
+type Tester = {
+   test: (value: any) => boolean;
+   stringify: () => string;
+};
 
+export class TypeMatcher<ExpectedType extends Object> {
+
+   private _testers: Array<Tester> = [];
    private _type: new (...args: Array<any>) => Object;
    public get type() {
       return this._type;
    }
 
-   public constructor(type: new (...args: Array<any>) => Object) {
+   public constructor(type: new (...args: Array<any>) => ExpectedType) {
       if (type === null || type === undefined) {
          throw new TypeError("type must not be null or undefined");
       }
 
       this._type = type;
+
+      this._testers.push({
+         stringify: () => `Any ${(this.type as any).name}`,
+         test: (value: any) => {
+            if ((<any> type) === String) {
+               return typeof value === "string" || value instanceof this._type;
+            }
+            else if ((<any> type) === Number) {
+               return typeof value === "number" || value instanceof this._type;
+            }
+            else if ((<any> type) === Boolean) {
+               return typeof value === "boolean" || value instanceof this._type;
+            }
+            else {
+               return value instanceof this._type;
+            }
+         }
+      });
    }
 
    public test(value: any) {
-      if (this._type === String) {
-         return typeof value === "string" || value instanceof this._type;
+      return this._testers.every(tester => tester.test(value));
+   }
+
+   public stringify() {
+      return this._testers.map(tester => tester.stringify()).join(" and ");
+   }
+
+   public thatMatches(key: string, value: any): this;
+   public thatMatches(properties: Object): this;
+   public thatMatches(delegate: (argument: ExpectedType) => boolean): this;
+   public thatMatches(first: string | Object | ((argument: ExpectedType) => boolean), second?: any): this {
+      if (null === first || undefined === first ) {
+         throw new TypeError("thatMatches requires none-null or non-undefined argument");
       }
-      else if (this._type === Number) {
-         return typeof value === "number" || value instanceof this._type;
+
+      if (typeof first === "string") {
+         return this.matchesKeyAndValue(first, second);
       }
-      else if (this._type === Boolean) {
-         return typeof value === "boolean" || value instanceof this._type;
+
+      if (typeof first === "function") {
+         return this.matchesDelegate(first);
       }
-      else {
-         return value instanceof this._type;
+
+      if (typeof first === "object") {
+         return this.matchesObjectLiteral(first);
       }
+
+      throw new Error("Invalid arguments");
+   }
+
+   private matchesKeyAndValue(key: string, value: any): this {
```

Comment:
If we could add underscores before these private functions that'd be perfect! :)